<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.30">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Directive for creating a script for your notebook – ssi_analysis_result_parsers</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-de070a7b0ab54f8780927367ac907214.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-0c9eb0277ab99f059f9e902403bd7a3d.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="Directive for creating a script for your notebook – ssi_analysis_result_parsers">
<meta property="og:description" content="TODO">
<meta property="og:site_name" content="ssi_analysis_result_parsers">
<meta name="twitter:title" content="Directive for creating a script for your notebook – ssi_analysis_result_parsers">
<meta name="twitter:description" content="TODO">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">ssi_analysis_result_parsers</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./blast_parser.html">Directive for creating a script for your notebook</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ssi_analysis_result_parsers</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./core.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Global static vars</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./hello_world.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Additional Learning Resources</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./template.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Directive for creating a script for your notebook</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./blast_parser.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Directive for creating a script for your notebook</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spyogenes_parser.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Directive for creating a script for your notebook</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./nmeningitidis_parser.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Directive for creating a script for your notebook</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./hinfluenzae_parser.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Directive for creating a script for your notebook</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./legionella_parser.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Directive for creating a script for your notebook</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ecoli_parser.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Directive for creating a script for your notebook</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#libraries" id="toc-libraries" class="nav-link active" data-scroll-target="#libraries">Libraries</a></li>
  <li><a href="#functions" id="toc-functions" class="nav-link" data-scroll-target="#functions">Functions</a>
  <ul class="collapse">
  <li><a href="#extract_allele_matches" id="toc-extract_allele_matches" class="nav-link" data-scroll-target="#extract_allele_matches">extract_allele_matches</a></li>
  <li><a href="#extract_presence_absence" id="toc-extract_presence_absence" class="nav-link" data-scroll-target="#extract_presence_absence">extract_presence_absence</a></li>
  <li><a href="#testing" id="toc-testing" class="nav-link" data-scroll-target="#testing">TESTING</a>
  <ul class="collapse">
  <li><a href="#allele_matches" id="toc-allele_matches" class="nav-link" data-scroll-target="#allele_matches">allele_matches</a></li>
  <li><a href="#presence_absence" id="toc-presence_absence" class="nav-link" data-scroll-target="#presence_absence">presence_absence</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#directive-for-ensuring-that-the-code-in-your-notebook-get-executed-as-a-script" id="toc-directive-for-ensuring-that-the-code-in-your-notebook-get-executed-as-a-script" class="nav-link" data-scroll-target="#directive-for-ensuring-that-the-code-in-your-notebook-get-executed-as-a-script">Directive for ensuring that the code in your notebook get executed as a script</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/thej-ssi/ssi_analysis_result_parsers/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div><div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="blast_parser.html.md"><i class="bi bi-file-code"></i>CommonMark</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Directive for creating a script for your notebook</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<section id="libraries" class="level1">
<h1>Libraries</h1>
<p>Include all the libraries which should be used in this module. You can also import modules from other notebooks; here, we have imported the functions in the core notebook.</p>
<div id="cell-3" class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This block should never be exported. It is to have python running in the project (and not the nbs) dir, and to initiate the package using pip.</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>os.chdir(core.PROJECT_DIR)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="functions" class="level1">
<h1>Functions</h1>
<p>Add your code here below. If your notebook will be used as a console-script, you need to add a “cli”-function, at the end (see <a href="https://dksund.sharepoint.com/:fl:/g/contentstorage/CSP_7c761ee7-b577-4e08-8517-bc82392bf65e/ETlSfUyArSNJhX8veMI_JQ8By1aXGHzDJkhotpfpXx4mmw?e=037EwH&amp;nav=cz0lMkZjb250ZW50c3RvcmFnZSUyRkNTUF83Yzc2MWVlNy1iNTc3LTRlMDgtODUxNy1iYzgyMzkyYmY2NWUmZD1iJTIxNXg1MmZIZTFDRTZGRjd5Q09TdjJYblkwVlNiWXFYcE1yaHVrVmZqTVJUVEE4X1VwZjhTd1JxcjRNdmFrSmh2RCZmPTAxVlVLVzVWSlpLSjZVWkFGTkVORVlLN1pQUERCRDZKSVAmYz0lMkYmYT1Mb29wQXBwJnA9JTQwZmx1aWR4JTJGbG9vcC1wYWdlLWNvbnRhaW5lciZ4PSU3QiUyMnclMjIlM0ElMjJUMFJUVUh4a2EzTjFibVF1YzJoaGNtVndiMmx1ZEM1amIyMThZaUUxZURVeVpraGxNVU5GTmtaR04zbERUMU4yTWxodVdUQldVMkpaY1Zod1RYSm9kV3RXWm1wTlVsUlVRVGhmVlhCbU9GTjNVbkZ5TkUxMllXdEthSFpFZkRBeFZsVkxWelZXU1RJMVJsaFBNalkyUlZkQ1FqTTFRVmhKVTBkRFVVcFdXa1klM0QlMjIlMkMlMjJpJTIyJTNBJTIyNzRmNzM1ZmUtYzg4Ny00MjhhLWFkZmYtNTEyZTg2YmNmZmQzJTIyJTdE">Coding in NBdev</a> (<strong>Code execution</strong> and <strong>Input, output and options</strong>) on loop for more details)</p>
<hr>
<p><a href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/blast_parser.py#L111" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="extract_allele_matches" class="level3">
<h3 class="anchored" data-anchor-id="extract_allele_matches">extract_allele_matches</h3>
<blockquote class="blockquote">
<pre><code> extract_allele_matches (blast_output_tsv:pathlib.Path, tsv_header:str,
                         include_match_stats=False)</code></pre>
</blockquote>
<p><em>Parse blast output tsv for best matching alleles returns: if include_match stats: { <gene_name_1>: <allele_number><strong><pident></pident></strong><plen>, <gene_name_2>: <allele_number><strong><pident></pident></strong><plen>, <gene_name_3>: <allele_number><strong><pident></pident></strong><plen> …} else: a dictionary (allele_dict) in the format { <gene_name_1>: <allele_number>, <gene_name_2>: <allele_number>, <gene_name_3>: <allele_number> …}</allele_number></gene_name_3></allele_number></gene_name_2></allele_number></gene_name_1></plen></allele_number></gene_name_3></plen></allele_number></gene_name_2></plen></allele_number></gene_name_1></em></p>
<hr>
<p><a href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/blast_parser.py#L33" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="extract_presence_absence" class="level3">
<h3 class="anchored" data-anchor-id="extract_presence_absence">extract_presence_absence</h3>
<blockquote class="blockquote">
<pre><code> extract_presence_absence (blast_output_tsv:pathlib.Path,
                           tsv_header:str='qseqid sseqid pident length
                           qlen qstart qend sstart send sseq evalue
                           bitscore', hits_as_string:bool=True,
                           include_match_stats=False,
                           pident_threshold:float=90,
                           plen_threshold:float=60, gene_names:list=None)</code></pre>
</blockquote>
<p><em>Parse blast output tsv for best matching alleles returns: if include_match stats: { <gene_name_1>: <allele_number>, <gene_name_2>: <allele_number>, <gene_name_3>: <allele_number> …} else: a dictionary (allele_dict) in the format { <gene_name_1>: <allele_number>, <gene_name_2>: <allele_number>, <gene_name_3>: <allele_number> …}</allele_number></gene_name_3></allele_number></gene_name_2></allele_number></gene_name_1></allele_number></gene_name_3></allele_number></gene_name_2></allele_number></gene_name_1></em></p>
</section>
<section id="testing" class="level2">
<h2 class="anchored" data-anchor-id="testing">TESTING</h2>
<div id="cell-8" class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>gene_presence_dict <span class="op">=</span> extract_presence_absence(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/blast_parser/gene_presence_absence_test.tsv"</span>,</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>                                     tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>                                     include_match_stats<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>                                     hits_as_string<span class="op">=</span><span class="va">True</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>                                     )</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(<span class="bu">len</span>(gene_presence_dict) <span class="op">==</span> <span class="dv">1</span>)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict[<span class="st">"genes_found"</span>] <span class="op">==</span> <span class="st">"asd, proA"</span>)</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>gene_presence_dict <span class="op">=</span> extract_presence_absence(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/blast_parser/gene_presence_absence_test.tsv"</span>,</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>                                     tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>                                     include_match_stats<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>                                     hits_as_string<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>                                     )</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict[<span class="st">"genes_found"</span>] <span class="op">==</span> <span class="st">"asd__98.732__100.0, proA__97.284__100.0"</span>)</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>gene_presence_dict <span class="op">=</span> extract_presence_absence(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/blast_parser/gene_presence_absence_test.tsv"</span>,</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>                                     tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>                                     include_match_stats<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>                                     hits_as_string<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>                                     pident_threshold<span class="op">=</span><span class="dv">80</span>,</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>                                     )</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict[<span class="st">"genes_found"</span>] <span class="op">==</span> <span class="st">"asd__98.732__100.0, pilE__82.835__100.0, proA__97.284__100.0"</span>)</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>gene_presence_dict <span class="op">=</span> extract_presence_absence(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/blast_parser/gene_presence_absence_test.tsv"</span>,</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>                                     tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>                                     include_match_stats<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>                                     hits_as_string<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>                                     gene_names <span class="op">=</span> [<span class="st">"asd"</span>,<span class="st">"proA"</span>,<span class="st">"pilE"</span>])</span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict[<span class="st">"asd"</span>] <span class="op">==</span> <span class="st">"1"</span>)</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict[<span class="st">"proA"</span>] <span class="op">==</span> <span class="st">"1"</span>)</span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict[<span class="st">"pilE"</span>] <span class="op">==</span> <span class="st">"0"</span>)</span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a>gene_presence_dict <span class="op">=</span> extract_presence_absence(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/blast_parser/gene_presence_absence_test.tsv"</span>,</span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a>                                     tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a>                                     include_match_stats<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a>                                     hits_as_string<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a>                                     gene_names <span class="op">=</span> [<span class="st">"asd"</span>,<span class="st">"proA"</span>,<span class="st">"pilE"</span>])</span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict[<span class="st">"asd"</span>] <span class="op">==</span> <span class="st">"98.732__100.0"</span>)</span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict[<span class="st">"proA"</span>] <span class="op">==</span> <span class="st">"97.284__100.0"</span>)</span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict[<span class="st">"pilE"</span>] <span class="op">==</span> <span class="st">"0"</span>)</span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a>gene_presence_dict <span class="op">=</span> extract_presence_absence(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/blast_parser/empty_gene_presence_absense_test.tsv"</span>,</span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a>                                     tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a>                                     include_match_stats<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a>                                     hits_as_string<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a>                                     gene_names<span class="op">=</span>[<span class="st">"lag-1"</span>]</span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a>                                     )</span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict[<span class="st">"lag-1"</span>] <span class="op">==</span> <span class="st">"0"</span>)</span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a>gene_presence_dict <span class="op">=</span> extract_presence_absence(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/empty_file.txt"</span>,</span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a>                                     tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a>                                     include_match_stats<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a>                                     hits_as_string<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb4-61"><a href="#cb4-61" aria-hidden="true" tabindex="-1"></a>                                     gene_names<span class="op">=</span>[<span class="st">"lag-1"</span>]</span>
<span id="cb4-62"><a href="#cb4-62" aria-hidden="true" tabindex="-1"></a>                                     )</span>
<span id="cb4-63"><a href="#cb4-63" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict[<span class="st">"lag-1"</span>] <span class="op">==</span> <span class="st">"0"</span>)</span>
<span id="cb4-64"><a href="#cb4-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-65"><a href="#cb4-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-66"><a href="#cb4-66" aria-hidden="true" tabindex="-1"></a>gene_presence_dict <span class="op">=</span> extract_presence_absence(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/empty_file.txt"</span>,</span>
<span id="cb4-67"><a href="#cb4-67" aria-hidden="true" tabindex="-1"></a>                                     tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-68"><a href="#cb4-68" aria-hidden="true" tabindex="-1"></a>                                     include_match_stats<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb4-69"><a href="#cb4-69" aria-hidden="true" tabindex="-1"></a>                                     hits_as_string<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb4-70"><a href="#cb4-70" aria-hidden="true" tabindex="-1"></a>                                     gene_names<span class="op">=</span>[<span class="st">"lag-1"</span>]</span>
<span id="cb4-71"><a href="#cb4-71" aria-hidden="true" tabindex="-1"></a>                                     )</span>
<span id="cb4-72"><a href="#cb4-72" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict[<span class="st">"lag-1"</span>] <span class="op">==</span> <span class="st">"0"</span>)</span>
<span id="cb4-73"><a href="#cb4-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-74"><a href="#cb4-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-75"><a href="#cb4-75" aria-hidden="true" tabindex="-1"></a>gene_presence_dict <span class="op">=</span> extract_presence_absence(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/empty_file.txt"</span>,</span>
<span id="cb4-76"><a href="#cb4-76" aria-hidden="true" tabindex="-1"></a>                                     tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-77"><a href="#cb4-77" aria-hidden="true" tabindex="-1"></a>                                     include_match_stats<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb4-78"><a href="#cb4-78" aria-hidden="true" tabindex="-1"></a>                                     hits_as_string<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb4-79"><a href="#cb4-79" aria-hidden="true" tabindex="-1"></a>                                     gene_names<span class="op">=</span>[<span class="st">"lag-1"</span>]</span>
<span id="cb4-80"><a href="#cb4-80" aria-hidden="true" tabindex="-1"></a>                                     )</span>
<span id="cb4-81"><a href="#cb4-81" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict[<span class="st">"genes_found"</span>] <span class="op">==</span> <span class="st">""</span>)</span>
<span id="cb4-82"><a href="#cb4-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-83"><a href="#cb4-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-84"><a href="#cb4-84" aria-hidden="true" tabindex="-1"></a>gene_presence_dict <span class="op">=</span> extract_presence_absence(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/empty_file.txt"</span>,</span>
<span id="cb4-85"><a href="#cb4-85" aria-hidden="true" tabindex="-1"></a>                                     tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-86"><a href="#cb4-86" aria-hidden="true" tabindex="-1"></a>                                     include_match_stats<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb4-87"><a href="#cb4-87" aria-hidden="true" tabindex="-1"></a>                                     hits_as_string<span class="op">=</span><span class="va">True</span></span>
<span id="cb4-88"><a href="#cb4-88" aria-hidden="true" tabindex="-1"></a>                                     )</span>
<span id="cb4-89"><a href="#cb4-89" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict[<span class="st">"genes_found"</span>] <span class="op">==</span> <span class="st">""</span>)</span>
<span id="cb4-90"><a href="#cb4-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-91"><a href="#cb4-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-92"><a href="#cb4-92" aria-hidden="true" tabindex="-1"></a><span class="co"># Test for missing file handling (should return None)</span></span>
<span id="cb4-93"><a href="#cb4-93" aria-hidden="true" tabindex="-1"></a>gene_presence_dict <span class="op">=</span> extract_presence_absence(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/file/that/does/not/exist.tsv"</span>,</span>
<span id="cb4-94"><a href="#cb4-94" aria-hidden="true" tabindex="-1"></a>                                     tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-95"><a href="#cb4-95" aria-hidden="true" tabindex="-1"></a>                                     include_match_stats<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb4-96"><a href="#cb4-96" aria-hidden="true" tabindex="-1"></a>                                     hits_as_string<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb4-97"><a href="#cb4-97" aria-hidden="true" tabindex="-1"></a>                                     gene_names<span class="op">=</span>[<span class="st">"lag-1"</span>]</span>
<span id="cb4-98"><a href="#cb4-98" aria-hidden="true" tabindex="-1"></a>                                     )</span>
<span id="cb4-99"><a href="#cb4-99" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict <span class="kw">is</span> <span class="va">None</span>)</span>
<span id="cb4-100"><a href="#cb4-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-101"><a href="#cb4-101" aria-hidden="true" tabindex="-1"></a><span class="co"># Check for incorrect number of columns handling (should return None)</span></span>
<span id="cb4-102"><a href="#cb4-102" aria-hidden="true" tabindex="-1"></a>gene_presence_dict <span class="op">=</span> extract_presence_absence(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/Legionella/test.sbt.tsv"</span>,</span>
<span id="cb4-103"><a href="#cb4-103" aria-hidden="true" tabindex="-1"></a>                                     tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-104"><a href="#cb4-104" aria-hidden="true" tabindex="-1"></a>                                     include_match_stats<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb4-105"><a href="#cb4-105" aria-hidden="true" tabindex="-1"></a>                                     hits_as_string<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb4-106"><a href="#cb4-106" aria-hidden="true" tabindex="-1"></a>                                     gene_names<span class="op">=</span>[<span class="st">"lag-1"</span>]</span>
<span id="cb4-107"><a href="#cb4-107" aria-hidden="true" tabindex="-1"></a>                                     )</span>
<span id="cb4-108"><a href="#cb4-108" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(gene_presence_dict <span class="kw">is</span> <span class="va">None</span>)</span>
<span id="cb4-109"><a href="#cb4-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-110"><a href="#cb4-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-111"><a href="#cb4-111" aria-hidden="true" tabindex="-1"></a>allele_dict <span class="op">=</span> extract_allele_matches(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/blast_parser/allele_matches_test.tsv"</span>,</span>
<span id="cb4-112"><a href="#cb4-112" aria-hidden="true" tabindex="-1"></a>                                     tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>)</span>
<span id="cb4-113"><a href="#cb4-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-114"><a href="#cb4-114" aria-hidden="true" tabindex="-1"></a>detailed_allele_dict <span class="op">=</span> extract_allele_matches(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/blast_parser/allele_matches_test.tsv"</span>,</span>
<span id="cb4-115"><a href="#cb4-115" aria-hidden="true" tabindex="-1"></a>                                     tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-116"><a href="#cb4-116" aria-hidden="true" tabindex="-1"></a>                                     include_match_stats<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb4-117"><a href="#cb4-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-118"><a href="#cb4-118" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(<span class="bu">len</span>(allele_dict) <span class="op">==</span> <span class="dv">7</span>)</span>
<span id="cb4-119"><a href="#cb4-119" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(allele_dict[<span class="st">'asd'</span>] <span class="op">==</span> <span class="st">"9"</span>)</span>
<span id="cb4-120"><a href="#cb4-120" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(detailed_allele_dict[<span class="st">'pilE'</span>] <span class="op">==</span> <span class="st">"3__100.0__100.0"</span>)</span>
<span id="cb4-121"><a href="#cb4-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-122"><a href="#cb4-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-123"><a href="#cb4-123" aria-hidden="true" tabindex="-1"></a><span class="co"># Test for missing file handling (should return None)</span></span>
<span id="cb4-124"><a href="#cb4-124" aria-hidden="true" tabindex="-1"></a>allele_dict <span class="op">=</span> extract_allele_matches(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/file/that/does/not/exist.tsv"</span>,</span>
<span id="cb4-125"><a href="#cb4-125" aria-hidden="true" tabindex="-1"></a>                                            tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-126"><a href="#cb4-126" aria-hidden="true" tabindex="-1"></a>                                            include_match_stats<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb4-127"><a href="#cb4-127" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(allele_dict <span class="kw">is</span> <span class="va">None</span>)</span>
<span id="cb4-128"><a href="#cb4-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-129"><a href="#cb4-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-130"><a href="#cb4-130" aria-hidden="true" tabindex="-1"></a><span class="co"># Test for incorrect number of columns handling (should return None)</span></span>
<span id="cb4-131"><a href="#cb4-131" aria-hidden="true" tabindex="-1"></a>allele_dict <span class="op">=</span> extract_allele_matches(blast_output_tsv<span class="op">=</span><span class="st">"./test_input/Legionella/test.sbt.tsv"</span>,</span>
<span id="cb4-132"><a href="#cb4-132" aria-hidden="true" tabindex="-1"></a>                                            tsv_header<span class="op">=</span><span class="st">"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore"</span>,</span>
<span id="cb4-133"><a href="#cb4-133" aria-hidden="true" tabindex="-1"></a>                                            include_match_stats<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb4-134"><a href="#cb4-134" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(allele_dict <span class="kw">is</span> <span class="va">None</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Blast output file ./test_input/blast_parser/empty_gene_presence_absense_test.tsv empty. Assuming 0 blast hits.
Blast output file ./test_input/empty_file.txt empty. Assuming 0 blast hits.
Blast output file ./test_input/empty_file.txt empty. Assuming 0 blast hits.
Blast output file ./test_input/empty_file.txt empty. Assuming 0 blast hits.
Blast output file ./test_input/empty_file.txt empty. Assuming 0 blast hits.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>No blast output found at ./test_input/file/that/does/not/exist.tsv
Failed to parse ./test_input/Legionella/test.sbt.tsv. Number of columns do not match length of provided header string
No blast output found at ./test_input/file/that/does/not/exist.tsv
Failed to parse ./test_input/Legionella/test.sbt.tsv. Number of columns do not match length of provided header string</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/blast_parser.py#L194" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="allele_matches" class="level3">
<h3 class="anchored" data-anchor-id="allele_matches">allele_matches</h3>
<blockquote class="blockquote">
<pre><code> allele_matches (blast_output:pathlib.Path=None,
                 blast_tsv_header:str='qseqid sseqid pident length qlen
                 qstart qend sstart send sseq evalue bitscore',
                 include_match_stats:bool=False,
                 output_file:pathlib.Path=None, config_file:str=None)</code></pre>
</blockquote>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>blast_output</td>
<td>Path</td>
<td>None</td>
<td>Path to blast output file. Generated with –outfmt 6 option</td>
</tr>
<tr class="even">
<td>blast_tsv_header</td>
<td>str</td>
<td>qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore</td>
<td>headers in blast output</td>
</tr>
<tr class="odd">
<td>include_match_stats</td>
<td>bool</td>
<td>False</td>
<td>True to include percent identity and percent length in output, false to only include allele number</td>
</tr>
<tr class="even">
<td>output_file</td>
<td>Path</td>
<td>None</td>
<td></td>
</tr>
<tr class="odd">
<td>config_file</td>
<td>str</td>
<td>None</td>
<td>config file to set env vars from</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/blast_parser.py#L169" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="presence_absence" class="level3">
<h3 class="anchored" data-anchor-id="presence_absence">presence_absence</h3>
<blockquote class="blockquote">
<pre><code> presence_absence (blast_output:pathlib.Path=None,
                   blast_tsv_header:str='qseqid sseqid pident length qlen
                   qstart qend sstart send sseq evalue bitscore',
                   hits_as_string:bool=True,
                   include_match_stats:bool=False,
                   percent_identityt:float=90, percent_length:float=60,
                   gene_names:list=None, output_file:pathlib.Path=None,
                   config_file:str=None)</code></pre>
</blockquote>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>blast_output</td>
<td>Path</td>
<td>None</td>
<td>Path to blast output file. Generated with –outfmt 6 option</td>
</tr>
<tr class="even">
<td>blast_tsv_header</td>
<td>str</td>
<td>qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore</td>
<td>headers in blast output</td>
</tr>
<tr class="odd">
<td>hits_as_string</td>
<td>bool</td>
<td>True</td>
<td>True to print a comma separated list of found genes on a single line. False to return a key: value pair for each gene</td>
</tr>
<tr class="even">
<td>include_match_stats</td>
<td>bool</td>
<td>False</td>
<td>True to include percent identity and percent length in output, false to only include present/absent</td>
</tr>
<tr class="odd">
<td>percent_identityt</td>
<td>float</td>
<td>90</td>
<td>percent identity threshold for considering a gene present</td>
</tr>
<tr class="even">
<td>percent_length</td>
<td>float</td>
<td>60</td>
<td>percent length threshold for considering a gene present</td>
</tr>
<tr class="odd">
<td>gene_names</td>
<td>list</td>
<td>None</td>
<td>name of genes to look for when hits_as_string = False</td>
</tr>
<tr class="even">
<td>output_file</td>
<td>Path</td>
<td>None</td>
<td></td>
</tr>
<tr class="odd">
<td>config_file</td>
<td>str</td>
<td>None</td>
<td>config file to set env vars from</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</section>
</section>
</section>
<section id="directive-for-ensuring-that-the-code-in-your-notebook-get-executed-as-a-script" class="level1">
<h1>Directive for ensuring that the code in your notebook get executed as a script</h1>
<p>The code-block here below is required to ensure that the code in the notebook is also transferred to the module (script), otherwise it will just be a notebook. See <a href="https://dksund.sharepoint.com/:fl:/g/contentstorage/CSP_7c761ee7-b577-4e08-8517-bc82392bf65e/ETlSfUyArSNJhX8veMI_JQ8By1aXGHzDJkhotpfpXx4mmw?e=037EwH&amp;nav=cz0lMkZjb250ZW50c3RvcmFnZSUyRkNTUF83Yzc2MWVlNy1iNTc3LTRlMDgtODUxNy1iYzgyMzkyYmY2NWUmZD1iJTIxNXg1MmZIZTFDRTZGRjd5Q09TdjJYblkwVlNiWXFYcE1yaHVrVmZqTVJUVEE4X1VwZjhTd1JxcjRNdmFrSmh2RCZmPTAxVlVLVzVWSlpLSjZVWkFGTkVORVlLN1pQUERCRDZKSVAmYz0lMkYmYT1Mb29wQXBwJnA9JTQwZmx1aWR4JTJGbG9vcC1wYWdlLWNvbnRhaW5lciZ4PSU3QiUyMnclMjIlM0ElMjJUMFJUVUh4a2EzTjFibVF1YzJoaGNtVndiMmx1ZEM1amIyMThZaUUxZURVeVpraGxNVU5GTmtaR04zbERUMU4yTWxodVdUQldVMkpaY1Zod1RYSm9kV3RXWm1wTlVsUlVRVGhmVlhCbU9GTjNVbkZ5TkUxMllXdEthSFpFZkRBeFZsVkxWelZXU1RJMVJsaFBNalkyUlZkQ1FqTTFRVmhKVTBkRFVVcFdXa1klM0QlMjIlMkMlMjJpJTIyJTNBJTIyNzRmNzM1ZmUtYzg4Ny00MjhhLWFkZmYtNTEyZTg2YmNmZmQzJTIyJTdE">Coding in NBdev</a> (<strong>Writing your own notebooks</strong>) on loop for more details.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/thej-ssi\.github\.io\/ssi_analysis_result_parsers");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/thej-ssi/ssi_analysis_result_parsers/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>