# Directive for creating a script for your notebook


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->

# Libraries

Include all the libraries which should be used in this module. You can
also import modules from other notebooks; here, we have imported the
functions in the core notebook.

``` python
# This block should never be exported. It is to have python running in the project (and not the nbs) dir, and to initiate the package using pip.
os.chdir(core.PROJECT_DIR)
```

# Functions

Add your code here below. If your notebook will be used as a
console-script, you need to add a “cli”-function, at the end (see
[Coding in
NBdev](https://dksund.sharepoint.com/:fl:/g/contentstorage/CSP_7c761ee7-b577-4e08-8517-bc82392bf65e/ETlSfUyArSNJhX8veMI_JQ8By1aXGHzDJkhotpfpXx4mmw?e=037EwH&nav=cz0lMkZjb250ZW50c3RvcmFnZSUyRkNTUF83Yzc2MWVlNy1iNTc3LTRlMDgtODUxNy1iYzgyMzkyYmY2NWUmZD1iJTIxNXg1MmZIZTFDRTZGRjd5Q09TdjJYblkwVlNiWXFYcE1yaHVrVmZqTVJUVEE4X1VwZjhTd1JxcjRNdmFrSmh2RCZmPTAxVlVLVzVWSlpLSjZVWkFGTkVORVlLN1pQUERCRDZKSVAmYz0lMkYmYT1Mb29wQXBwJnA9JTQwZmx1aWR4JTJGbG9vcC1wYWdlLWNvbnRhaW5lciZ4PSU3QiUyMnclMjIlM0ElMjJUMFJUVUh4a2EzTjFibVF1YzJoaGNtVndiMmx1ZEM1amIyMThZaUUxZURVeVpraGxNVU5GTmtaR04zbERUMU4yTWxodVdUQldVMkpaY1Zod1RYSm9kV3RXWm1wTlVsUlVRVGhmVlhCbU9GTjNVbkZ5TkUxMllXdEthSFpFZkRBeFZsVkxWelZXU1RJMVJsaFBNalkyUlZkQ1FqTTFRVmhKVTBkRFVVcFdXa1klM0QlMjIlMkMlMjJpJTIyJTNBJTIyNzRmNzM1ZmUtYzg4Ny00MjhhLWFkZmYtNTEyZTg2YmNmZmQzJTIyJTdE)
(**Code execution** and **Input, output and options**) on loop for more
details)

------------------------------------------------------------------------

<a
href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/Hinfluenzae_parser.py#L167"
target="_blank" style="float:right; font-size:smaller">source</a>

### HinfluenzaeResults

>  HinfluenzaeResults (results_dict)

*Initialize self. See help(type(self)) for accurate signature.*

------------------------------------------------------------------------

<a
href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/Hinfluenzae_parser.py#L101"
target="_blank" style="float:right; font-size:smaller">source</a>

### extract_ariba_ftsI_snps

>  extract_ariba_ftsI_snps (ariba_output_tsv:pathlib.Path,
>                               ftsI_types_tsv:pathlib.Path)

------------------------------------------------------------------------

<a
href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/Hinfluenzae_parser.py#L75"
target="_blank" style="float:right; font-size:smaller">source</a>

### extract_hicap_results

>  extract_hicap_results (hicap_tsv:pathlib.Path)

------------------------------------------------------------------------

<a
href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/Hinfluenzae_parser.py#L40"
target="_blank" style="float:right; font-size:smaller">source</a>

### get_biotype_from_gene_presence

>  get_biotype_from_gene_presence (biotype_gene_presence_dict:pathlib.Path)

*Convert biotype gene presence dict to biotype*

## TESTING

------------------------------------------------------------------------

<a
href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/Hinfluenzae_parser.py#L294"
target="_blank" style="float:right; font-size:smaller">source</a>

### Hinfluenzae_batch_parser

>  Hinfluenzae_batch_parser (file_path_tsv:pathlib.Path=None,
>                                output_file:pathlib.Path=None)

<table>
<colgroup>
<col style="width: 6%" />
<col style="width: 25%" />
<col style="width: 34%" />
<col style="width: 34%" />
</colgroup>
<thead>
<tr>
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>file_path_tsv</td>
<td>Path</td>
<td>None</td>
<td>Path to tsv containing file paths to the outputs from tools to be
parsed. Must contain headers “sample_name”, “sbt_results”, and
“lag1_blast_results”</td>
</tr>
<tr>
<td>output_file</td>
<td>Path</td>
<td>None</td>
<td>Path to output tsv</td>
</tr>
<tr>
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>

------------------------------------------------------------------------

<a
href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/Hinfluenzae_parser.py#L274"
target="_blank" style="float:right; font-size:smaller">source</a>

### Hinfluenzae_parser

>  Hinfluenzae_parser (ftsI_ariba_tsv:pathlib.Path=None,
>                          hicap_tsv:pathlib.Path=None,
>                          biotype_blast_tsv:pathlib.Path=None,
>                          ftsI_types_tsv:pathlib.Path=None,
>                          output_file:pathlib.Path=None, sample_name:str=None)

<table>
<colgroup>
<col style="width: 6%" />
<col style="width: 25%" />
<col style="width: 34%" />
<col style="width: 34%" />
</colgroup>
<thead>
<tr>
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>ftsI_ariba_tsv</td>
<td>Path</td>
<td>None</td>
<td>Path to report.tsv from ftsI ariba output</td>
</tr>
<tr>
<td>hicap_tsv</td>
<td>Path</td>
<td>None</td>
<td>Path to hicap tsv output</td>
</tr>
<tr>
<td>biotype_blast_tsv</td>
<td>Path</td>
<td>None</td>
<td>Path to output from biotype gene blast. Generated with blastn -query
biotype_genes.fasta -subject assembly.fasta -outfmt “6 qseqid sseqid
pident length qlen qstart qend sstart send sseq evalue bitscore”</td>
</tr>
<tr>
<td>ftsI_types_tsv</td>
<td>Path</td>
<td>None</td>
<td>Path to table to convert ftsI snps to ftsI types</td>
</tr>
<tr>
<td>output_file</td>
<td>Path</td>
<td>None</td>
<td>Path to output tsv</td>
</tr>
<tr>
<td>sample_name</td>
<td>str</td>
<td>None</td>
<td></td>
</tr>
<tr>
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>

# Directive for ensuring that the code in your notebook get executed as a script

The code-block here below is required to ensure that the code in the
notebook is also transferred to the module (script), otherwise it will
just be a notebook. See [Coding in
NBdev](https://dksund.sharepoint.com/:fl:/g/contentstorage/CSP_7c761ee7-b577-4e08-8517-bc82392bf65e/ETlSfUyArSNJhX8veMI_JQ8By1aXGHzDJkhotpfpXx4mmw?e=037EwH&nav=cz0lMkZjb250ZW50c3RvcmFnZSUyRkNTUF83Yzc2MWVlNy1iNTc3LTRlMDgtODUxNy1iYzgyMzkyYmY2NWUmZD1iJTIxNXg1MmZIZTFDRTZGRjd5Q09TdjJYblkwVlNiWXFYcE1yaHVrVmZqTVJUVEE4X1VwZjhTd1JxcjRNdmFrSmh2RCZmPTAxVlVLVzVWSlpLSjZVWkFGTkVORVlLN1pQUERCRDZKSVAmYz0lMkYmYT1Mb29wQXBwJnA9JTQwZmx1aWR4JTJGbG9vcC1wYWdlLWNvbnRhaW5lciZ4PSU3QiUyMnclMjIlM0ElMjJUMFJUVUh4a2EzTjFibVF1YzJoaGNtVndiMmx1ZEM1amIyMThZaUUxZURVeVpraGxNVU5GTmtaR04zbERUMU4yTWxodVdUQldVMkpaY1Zod1RYSm9kV3RXWm1wTlVsUlVRVGhmVlhCbU9GTjNVbkZ5TkUxMllXdEthSFpFZkRBeFZsVkxWelZXU1RJMVJsaFBNalkyUlZkQ1FqTTFRVmhKVTBkRFVVcFdXa1klM0QlMjIlMkMlMjJpJTIyJTNBJTIyNzRmNzM1ZmUtYzg4Ny00MjhhLWFkZmYtNTEyZTg2YmNmZmQzJTIyJTdE)
(**Writing your own notebooks**) on loop for more details.
