# Directive for creating a script for your notebook


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->

# Libraries

Include all the libraries which should be used in this module. You can
also import modules from other notebooks; here, we have imported the
functions in the core notebook.

``` python
# This block should never be exported. It is to have python running in the project (and not the nbs) dir, and to initiate the package using pip.
os.chdir(core.PROJECT_DIR)
```

# Functions

Add your code here below. If your notebook will be used as a
console-script, you need to add a “cli”-function, at the end (see
[Coding in
NBdev](https://dksund.sharepoint.com/:fl:/g/contentstorage/CSP_7c761ee7-b577-4e08-8517-bc82392bf65e/ETlSfUyArSNJhX8veMI_JQ8By1aXGHzDJkhotpfpXx4mmw?e=037EwH&nav=cz0lMkZjb250ZW50c3RvcmFnZSUyRkNTUF83Yzc2MWVlNy1iNTc3LTRlMDgtODUxNy1iYzgyMzkyYmY2NWUmZD1iJTIxNXg1MmZIZTFDRTZGRjd5Q09TdjJYblkwVlNiWXFYcE1yaHVrVmZqTVJUVEE4X1VwZjhTd1JxcjRNdmFrSmh2RCZmPTAxVlVLVzVWSlpLSjZVWkFGTkVORVlLN1pQUERCRDZKSVAmYz0lMkYmYT1Mb29wQXBwJnA9JTQwZmx1aWR4JTJGbG9vcC1wYWdlLWNvbnRhaW5lciZ4PSU3QiUyMnclMjIlM0ElMjJUMFJUVUh4a2EzTjFibVF1YzJoaGNtVndiMmx1ZEM1amIyMThZaUUxZURVeVpraGxNVU5GTmtaR04zbERUMU4yTWxodVdUQldVMkpaY1Zod1RYSm9kV3RXWm1wTlVsUlVRVGhmVlhCbU9GTjNVbkZ5TkUxMllXdEthSFpFZkRBeFZsVkxWelZXU1RJMVJsaFBNalkyUlZkQ1FqTTFRVmhKVTBkRFVVcFdXa1klM0QlMjIlMkMlMjJpJTIyJTNBJTIyNzRmNzM1ZmUtYzg4Ny00MjhhLWFkZmYtNTEyZTg2YmNmZmQzJTIyJTdE)
(**Code execution** and **Input, output and options**) on loop for more
details)

------------------------------------------------------------------------

<a
href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/Legionella_parser.py#L65"
target="_blank" style="float:right; font-size:smaller">source</a>

### LegionellaResults

>  LegionellaResults (results_dict)

*Initialize self. See help(type(self)) for accurate signature.*

------------------------------------------------------------------------

<a
href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/Legionella_parser.py#L39"
target="_blank" style="float:right; font-size:smaller">source</a>

### extract_legionella_sbt

>  extract_legionella_sbt (legionella_sbt_results_tsv:pathlib.Path)

*Extract results from Legionella SBT results file Returns a dictionary
with ST, allele variant for each gene, and notes from output*

## TESTING

``` python
f = LegionellaResults.from_results_tsv("./test_output/test_batch_output.tsv")
assert(len(f) == 2)
assert(f.results_dict["sample_1"]["ST"] == 23)


f = LegionellaResults.from_tool_paths_dict(file_paths=  {"sample_1": {"sbt_results": "test_input/Legionella/test.sbt.tsv", "lag1_blast_results": "test_input/Legionella/lag-1_blast.tsv"},
                                                            "sample_2": {"sbt_results": "test_input/Legionella/test2.sbt.tsv", "lag1_blast_results": "test_input/Legionella/lag-1_blast_2.tsv"}})



f = LegionellaResults.from_tool_paths(legionella_sbt_results_tsv="test_input/Legionella/test.sbt.tsv",
                                      lag1_blast_tsv="test_input/Legionella/lag-1_blast.tsv")



f = LegionellaResults.from_tool_paths_tsv(tool_paths_tsv="test_input/Legionella/batch_parser_file_paths.tsv")

assert(len(f) == 4)
assert(len(f.results_df) == 4)
assert(len(f.results_df.columns) == 10)
```

    Blast output file test_input/Legionella/lag-1_blast_2.tsv empty. Assuming 0 blast hits.
    {'ST': 23, 'flaA': 2, 'pilE': 3, 'asd': 9, 'mip': 10, 'mompS': 2, 'proA': 1, 'neuA': 6, 'notes': 'Exact ST match, Heterozygous mompS alleles, High confidence mompS allele call', 'lag-1': '1'}
    Blast output file test_input/Legionella/lag-1_blast_2.tsv empty. Assuming 0 blast hits.
    {'ST': 182, 'flaA': 3, 'pilE': 4, 'asd': 1, 'mip': 3, 'mompS': 35, 'proA': 9, 'neuA': 11, 'notes': 'Exact ST match, Heterozygous mompS alleles, High confidence mompS allele call', 'lag-1': '0'}
    Blast output file test_input/empty_file.txt empty. Assuming 0 blast hits.
    {'ST': 182, 'flaA': 3, 'pilE': 4, 'asd': 1, 'mip': 3, 'mompS': 35, 'proA': 9, 'neuA': 11, 'notes': 'Exact ST match, Heterozygous mompS alleles, High confidence mompS allele call', 'lag-1': '0'}
    Blast output file test_input/Legionella/lag-1_blast_2.tsv empty. Assuming 0 blast hits.
    {'lag-1': '0'}

    Legionella SBT output empty at test_input/empty_file.txt

------------------------------------------------------------------------

<a
href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/Legionella_parser.py#L163"
target="_blank" style="float:right; font-size:smaller">source</a>

### legionella_batch_parser

>  legionella_batch_parser (file_path_tsv:pathlib.Path=None,
>                               output_file:pathlib.Path=None)

<table>
<colgroup>
<col style="width: 6%" />
<col style="width: 25%" />
<col style="width: 34%" />
<col style="width: 34%" />
</colgroup>
<thead>
<tr>
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>file_path_tsv</td>
<td>Path</td>
<td>None</td>
<td>Path to tsv containing file paths to the outputs from tools to be
parsed. Must contain headers “sample_name”, “sbt_results”, and
“lag1_blast_results”</td>
</tr>
<tr>
<td>output_file</td>
<td>Path</td>
<td>None</td>
<td>Path to output tsv</td>
</tr>
<tr>
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>

------------------------------------------------------------------------

<a
href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/Legionella_parser.py#L147"
target="_blank" style="float:right; font-size:smaller">source</a>

### legionella_parser

>  legionella_parser (legionella_sbt_file:pathlib.Path=None,
>                         lag_1_blast_output:pathlib.Path=None,
>                         output_file:pathlib.Path=None, sample_name:str=None)

<table>
<colgroup>
<col style="width: 6%" />
<col style="width: 25%" />
<col style="width: 34%" />
<col style="width: 34%" />
</colgroup>
<thead>
<tr>
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>legionella_sbt_file</td>
<td>Path</td>
<td>None</td>
<td>Path “*.sbt.tsv from legionella_sbt program”</td>
</tr>
<tr>
<td>lag_1_blast_output</td>
<td>Path</td>
<td>None</td>
<td>Path to output from lag1_blast. Generated with blastn -query
lag-1.fasta -subject assembly.fasta -outfmt “6 qseqid sseqid pident
length qlen qstart qend sstart send sseq evalue bitscore”</td>
</tr>
<tr>
<td>output_file</td>
<td>Path</td>
<td>None</td>
<td>Path to output tsv</td>
</tr>
<tr>
<td>sample_name</td>
<td>str</td>
<td>None</td>
<td></td>
</tr>
<tr>
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>

# Directive for ensuring that the code in your notebook get executed as a script

The code-block here below is required to ensure that the code in the
notebook is also transferred to the module (script), otherwise it will
just be a notebook. See [Coding in
NBdev](https://dksund.sharepoint.com/:fl:/g/contentstorage/CSP_7c761ee7-b577-4e08-8517-bc82392bf65e/ETlSfUyArSNJhX8veMI_JQ8By1aXGHzDJkhotpfpXx4mmw?e=037EwH&nav=cz0lMkZjb250ZW50c3RvcmFnZSUyRkNTUF83Yzc2MWVlNy1iNTc3LTRlMDgtODUxNy1iYzgyMzkyYmY2NWUmZD1iJTIxNXg1MmZIZTFDRTZGRjd5Q09TdjJYblkwVlNiWXFYcE1yaHVrVmZqTVJUVEE4X1VwZjhTd1JxcjRNdmFrSmh2RCZmPTAxVlVLVzVWSlpLSjZVWkFGTkVORVlLN1pQUERCRDZKSVAmYz0lMkYmYT1Mb29wQXBwJnA9JTQwZmx1aWR4JTJGbG9vcC1wYWdlLWNvbnRhaW5lciZ4PSU3QiUyMnclMjIlM0ElMjJUMFJUVUh4a2EzTjFibVF1YzJoaGNtVndiMmx1ZEM1amIyMThZaUUxZURVeVpraGxNVU5GTmtaR04zbERUMU4yTWxodVdUQldVMkpaY1Zod1RYSm9kV3RXWm1wTlVsUlVRVGhmVlhCbU9GTjNVbkZ5TkUxMllXdEthSFpFZkRBeFZsVkxWelZXU1RJMVJsaFBNalkyUlZkQ1FqTTFRVmhKVTBkRFVVcFdXa1klM0QlMjIlMkMlMjJpJTIyJTNBJTIyNzRmNzM1ZmUtYzg4Ny00MjhhLWFkZmYtNTEyZTg2YmNmZmQzJTIyJTdE)
(**Writing your own notebooks**) on loop for more details.
