# Directive for creating a script for your notebook


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->

# Libraries

Include all the libraries which should be used in this module. You can
also import modules from other notebooks; here, we have imported the
functions in the core notebook.

``` python
# This block should never be exported. It is to have python running in the project (and not the nbs) dir, and to initiate the package using pip.
os.chdir(core.PROJECT_DIR)
```

# Functions

Add your code here below. If your notebook will be used as a
console-script, you need to add a “cli”-function, at the end (see
[Coding in
NBdev](https://dksund.sharepoint.com/:fl:/g/contentstorage/CSP_7c761ee7-b577-4e08-8517-bc82392bf65e/ETlSfUyArSNJhX8veMI_JQ8By1aXGHzDJkhotpfpXx4mmw?e=037EwH&nav=cz0lMkZjb250ZW50c3RvcmFnZSUyRkNTUF83Yzc2MWVlNy1iNTc3LTRlMDgtODUxNy1iYzgyMzkyYmY2NWUmZD1iJTIxNXg1MmZIZTFDRTZGRjd5Q09TdjJYblkwVlNiWXFYcE1yaHVrVmZqTVJUVEE4X1VwZjhTd1JxcjRNdmFrSmh2RCZmPTAxVlVLVzVWSlpLSjZVWkFGTkVORVlLN1pQUERCRDZKSVAmYz0lMkYmYT1Mb29wQXBwJnA9JTQwZmx1aWR4JTJGbG9vcC1wYWdlLWNvbnRhaW5lciZ4PSU3QiUyMnclMjIlM0ElMjJUMFJUVUh4a2EzTjFibVF1YzJoaGNtVndiMmx1ZEM1amIyMThZaUUxZURVeVpraGxNVU5GTmtaR04zbERUMU4yTWxodVdUQldVMkpaY1Zod1RYSm9kV3RXWm1wTlVsUlVRVGhmVlhCbU9GTjNVbkZ5TkUxMllXdEthSFpFZkRBeFZsVkxWelZXU1RJMVJsaFBNalkyUlZkQ1FqTTFRVmhKVTBkRFVVcFdXa1klM0QlMjIlMkMlMjJpJTIyJTNBJTIyNzRmNzM1ZmUtYzg4Ny00MjhhLWFkZmYtNTEyZTg2YmNmZmQzJTIyJTdE)
(**Code execution** and **Input, output and options**) on loop for more
details)

------------------------------------------------------------------------

<a
href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/Spyogenes_parser.py#L212"
target="_blank" style="float:right; font-size:smaller">source</a>

### SpyogenesResults

>  SpyogenesResults (results_dict)

*Initialize self. See help(type(self)) for accurate signature.*

------------------------------------------------------------------------

<a
href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/Spyogenes_parser.py#L39"
target="_blank" style="float:right; font-size:smaller">source</a>

### extract_emm_type

>  extract_emm_type (emm_blast_tsv:pathlib.Path)

\*with open(emm_cluster_file) as f: emm_clusters = {} for line in f:
line = line.rstrip(” “).split() for ele in line: emm_clusters\[ele\] =
line\[0\]

try: mga_blast_df = pandas.read_csv(mga_blast_tsv, sep=’ ‘, header =
None) mga_blast_df.columns = “qseqid sseqid pident length qlen qstart
qend sstart send evalue bitscore”.split(’ ’)
mga_blast_df.sort_values(by=\[‘bitscore’\], ascending= False).iloc\[0\]
mga_pos = mga_blast_df.iloc\[0\]\[‘sstart’\]

except pandas.errors.EmptyDataError: print(f”No mga matches found in
assembly”)\*

## TESTING

``` python
emm_typing_dict = extract_emm_type(emm_blast_tsv="test_input/Spyogenes/emm_typing/test1.emm.blast.tsv")

assert(emm_typing_dict["EMM_type"] == "EMM87.0")
assert(emm_typing_dict["ENN_type"] == "EMM159.0*")
assert(emm_typing_dict["MRP_type"] == "-")
assert(emm_typing_dict["emm_typing_notes"] == "ENN159.0 with pident 97.78 and length 180/180")



Spyogenes_results = SpyogenesResults.from_tool_paths(emm_blast_tsv="test_input/Spyogenes/emm_typing/test2.emm.blast.tsv", sample_name="test_sample")

assert(list(Spyogenes_results.results_dict.keys())[0] == "test_sample")

Spyogenes_results = SpyogenesResults.from_tool_paths_dict(file_paths=  {"sample_3": {"emm_results": "test_input/Spyogenes/emm_typing/test3.emm.blast.tsv"},
                                                                        "sample_4": {"emm_results": "test_input/Spyogenes/emm_typing/test4.emm.blast.tsv"}})

assert(Spyogenes_results.results_dict["sample_3"]["EMM_type"] == "EMM77.0")
assert(Spyogenes_results.results_dict["sample_3"]["ENN_type"] == "EMM159.0")
assert(Spyogenes_results.results_dict["sample_3"]["MRP_type"] == "EMM141.3*")
assert(Spyogenes_results.results_dict["sample_3"]["emm_typing_notes"] == "MRP141.3 with pident 97.22 and length 180/180")



assert(Spyogenes_results.results_df["EMM_type"]["sample_4"] == "EMM13.0")
assert(Spyogenes_results.results_df["ENN_type"]["sample_4"] == "EMM203.4*")
assert(Spyogenes_results.results_df["MRP_type"]["sample_4"] == "EMM141.4")
assert(Spyogenes_results.results_df["emm_typing_notes"]["sample_4"] == "ENN203.4 with pident 97.22 and length 180/180")


Spyogenes_results = SpyogenesResults.from_tool_paths_tsv(tool_paths_tsv="test_input/Spyogenes/batch_parser_file_paths.tsv")


assert(Spyogenes_results.results_dict["sample_5"]["EMM_type"] == "EMM81.0")
assert(Spyogenes_results.results_dict["sample_5"]["ENN_type"] == "-")
assert(Spyogenes_results.results_dict["sample_5"]["MRP_type"] == "EMM156.4*")
assert(Spyogenes_results.results_dict["sample_5"]["emm_typing_notes"] == "EMM156.4 with pident 99.44 and length 180/180, EMM redesignated due to known MRP+EMM operon")


assert(Spyogenes_results.results_df["EMM_type"]["sample_7"] == "-")
assert(Spyogenes_results.results_df["ENN_type"]["sample_7"] == "-")
assert(Spyogenes_results.results_df["MRP_type"]["sample_7"] == "-")
assert(Spyogenes_results.results_df["emm_typing_notes"]["sample_7"] == "EMM and EMM-like genes found on multiple contigs. Alleles found: 203.4*/28.0")

assert(Spyogenes_results.results_df["EMM_type"]["sample_empty"] == "-")
assert(Spyogenes_results.results_df["ENN_type"]["sample_empty"] == "-")
assert(Spyogenes_results.results_df["MRP_type"]["sample_empty"] == "-")
assert(Spyogenes_results.results_df["emm_typing_notes"]["sample_empty"] == "Empty blast output, no EMM genes detected")


assert(Spyogenes_results.results_df["EMM_type"]["sample_nonexist"] == "-")
assert(Spyogenes_results.results_df["ENN_type"]["sample_nonexist"] == "-")
assert(Spyogenes_results.results_df["MRP_type"]["sample_nonexist"] == "-")
assert(Spyogenes_results.results_df["emm_typing_notes"]["sample_nonexist"] == "No blast output found for EMM genes")
```

------------------------------------------------------------------------

<a
href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/Spyogenes_parser.py#L285"
target="_blank" style="float:right; font-size:smaller">source</a>

### Spyogenes_batch_parser

>  Spyogenes_batch_parser (file_path_tsv:pathlib.Path=None,
>                              output_file:pathlib.Path=None)

<table>
<colgroup>
<col style="width: 6%" />
<col style="width: 25%" />
<col style="width: 34%" />
<col style="width: 34%" />
</colgroup>
<thead>
<tr>
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>file_path_tsv</td>
<td>Path</td>
<td>None</td>
<td>Path to tsv containing file paths to the outputs from tools to be
parsed. Must contain headers “sample_name”, “sbt_results”, and
“lag1_blast_results”</td>
</tr>
<tr>
<td>output_file</td>
<td>Path</td>
<td>None</td>
<td>Path to output tsv</td>
</tr>
<tr>
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>

------------------------------------------------------------------------

<a
href="https://github.com/thej-ssi/ssi_analysis_result_parsers/blob/main/ssi_analysis_result_parsers/Spyogenes_parser.py#L272"
target="_blank" style="float:right; font-size:smaller">source</a>

### Spyogenes_parser

>  Spyogenes_parser (emm_blast_tsv:pathlib.Path=None,
>                        output_file:pathlib.Path=None, sample_name:str=None)

<table>
<colgroup>
<col style="width: 6%" />
<col style="width: 25%" />
<col style="width: 34%" />
<col style="width: 34%" />
</colgroup>
<thead>
<tr>
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>emm_blast_tsv</td>
<td>Path</td>
<td>None</td>
<td>Blast output from blasting EMM and emm-like genes</td>
</tr>
<tr>
<td>output_file</td>
<td>Path</td>
<td>None</td>
<td>Path to output tsv</td>
</tr>
<tr>
<td>sample_name</td>
<td>str</td>
<td>None</td>
<td></td>
</tr>
<tr>
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>

# Directive for ensuring that the code in your notebook get executed as a script

The code-block here below is required to ensure that the code in the
notebook is also transferred to the module (script), otherwise it will
just be a notebook. See [Coding in
NBdev](https://dksund.sharepoint.com/:fl:/g/contentstorage/CSP_7c761ee7-b577-4e08-8517-bc82392bf65e/ETlSfUyArSNJhX8veMI_JQ8By1aXGHzDJkhotpfpXx4mmw?e=037EwH&nav=cz0lMkZjb250ZW50c3RvcmFnZSUyRkNTUF83Yzc2MWVlNy1iNTc3LTRlMDgtODUxNy1iYzgyMzkyYmY2NWUmZD1iJTIxNXg1MmZIZTFDRTZGRjd5Q09TdjJYblkwVlNiWXFYcE1yaHVrVmZqTVJUVEE4X1VwZjhTd1JxcjRNdmFrSmh2RCZmPTAxVlVLVzVWSlpLSjZVWkFGTkVORVlLN1pQUERCRDZKSVAmYz0lMkYmYT1Mb29wQXBwJnA9JTQwZmx1aWR4JTJGbG9vcC1wYWdlLWNvbnRhaW5lciZ4PSU3QiUyMnclMjIlM0ElMjJUMFJUVUh4a2EzTjFibVF1YzJoaGNtVndiMmx1ZEM1amIyMThZaUUxZURVeVpraGxNVU5GTmtaR04zbERUMU4yTWxodVdUQldVMkpaY1Zod1RYSm9kV3RXWm1wTlVsUlVRVGhmVlhCbU9GTjNVbkZ5TkUxMllXdEthSFpFZkRBeFZsVkxWelZXU1RJMVJsaFBNalkyUlZkQ1FqTTFRVmhKVTBkRFVVcFdXa1klM0QlMjIlMkMlMjJpJTIyJTNBJTIyNzRmNzM1ZmUtYzg4Ny00MjhhLWFkZmYtNTEyZTg2YmNmZmQzJTIyJTdE)
(**Writing your own notebooks**) on loop for more details.
