{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Directive for creating a script for your notebook\n",
    "\n",
    "The block here below is required at the top of each notebook that you want to create a script for. You will also need to edit the \"settings.ini\" file, to create a script (see [Coding in NBdev](https://dksund.sharepoint.com/:fl:/g/contentstorage/CSP_7c761ee7-b577-4e08-8517-bc82392bf65e/ETlSfUyArSNJhX8veMI_JQ8By1aXGHzDJkhotpfpXx4mmw?e=037EwH&nav=cz0lMkZjb250ZW50c3RvcmFnZSUyRkNTUF83Yzc2MWVlNy1iNTc3LTRlMDgtODUxNy1iYzgyMzkyYmY2NWUmZD1iJTIxNXg1MmZIZTFDRTZGRjd5Q09TdjJYblkwVlNiWXFYcE1yaHVrVmZqTVJUVEE4X1VwZjhTd1JxcjRNdmFrSmh2RCZmPTAxVlVLVzVWSlpLSjZVWkFGTkVORVlLN1pQUERCRDZKSVAmYz0lMkYmYT1Mb29wQXBwJnA9JTQwZmx1aWR4JTJGbG9vcC1wYWdlLWNvbnRhaW5lciZ4PSU3QiUyMnclMjIlM0ElMjJUMFJUVUh4a2EzTjFibVF1YzJoaGNtVndiMmx1ZEM1amIyMThZaUUxZURVeVpraGxNVU5GTmtaR04zbERUMU4yTWxodVdUQldVMkpaY1Zod1RYSm9kV3RXWm1wTlVsUlVRVGhmVlhCbU9GTjNVbkZ5TkUxMllXdEthSFpFZkRBeFZsVkxWelZXU1RJMVJsaFBNalkyUlZkQ1FqTTFRVmhKVTBkRFVVcFdXa1klM0QlMjIlMkMlMjJpJTIyJTNBJTIyNzRmNzM1ZmUtYzg4Ny00MjhhLWFkZmYtNTEyZTg2YmNmZmQzJTIyJTdE) \n",
    "(**Writing your own notebooks**) on loop for more details). Replace **some_string** with a name that makes sense for your notebook. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# |default_exp blast_parser\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Libraries\n",
    "Include all the libraries which should be used in this module. You can also import modules from other notebooks; here, we have imported the functions in the core notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# |export\n",
    "\n",
    "# standard libs\n",
    "import os\n",
    "import re\n",
    "\n",
    "# Common to template\n",
    "# add into settings.ini, requirements, package name is python-dotenv, for conda build ensure `conda config --add channels conda-forge`\n",
    "import dotenv  # for loading config from .env files, https://pypi.org/project/python-dotenv/\n",
    "import envyaml  # Allows to loads env vars into a yaml file, https://github.com/thesimj/envyaml\n",
    "import fastcore  # To add functionality related to nbdev development, https://github.com/fastai/fastcore/\n",
    "from fastcore import (\n",
    "    test,\n",
    ")\n",
    "from fastcore.script import (\n",
    "    call_parse,\n",
    ")  # for @call_parse, https://fastcore.fast.ai/script\n",
    "import json  # for nicely printing json and yaml\n",
    "\n",
    "#import functions from core module (optional, but most likely needed). \n",
    "from ssi_analysis_result_parsers import(\n",
    "    core\n",
    ")\n",
    "\n",
    "# Project specific libraries\n",
    "from pathlib import Path\n",
    "import pandas\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This block should never be exported. It is to have python running in the project (and not the nbs) dir, and to initiate the package using pip.\n",
    "os.chdir(core.PROJECT_DIR)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions\n",
    "\n",
    "Add your code here below. If your notebook will be used as a console-script, you need to add a \"cli\"-function, at the end (see [Coding in NBdev](https://dksund.sharepoint.com/:fl:/g/contentstorage/CSP_7c761ee7-b577-4e08-8517-bc82392bf65e/ETlSfUyArSNJhX8veMI_JQ8By1aXGHzDJkhotpfpXx4mmw?e=037EwH&nav=cz0lMkZjb250ZW50c3RvcmFnZSUyRkNTUF83Yzc2MWVlNy1iNTc3LTRlMDgtODUxNy1iYzgyMzkyYmY2NWUmZD1iJTIxNXg1MmZIZTFDRTZGRjd5Q09TdjJYblkwVlNiWXFYcE1yaHVrVmZqTVJUVEE4X1VwZjhTd1JxcjRNdmFrSmh2RCZmPTAxVlVLVzVWSlpLSjZVWkFGTkVORVlLN1pQUERCRDZKSVAmYz0lMkYmYT1Mb29wQXBwJnA9JTQwZmx1aWR4JTJGbG9vcC1wYWdlLWNvbnRhaW5lciZ4PSU3QiUyMnclMjIlM0ElMjJUMFJUVUh4a2EzTjFibVF1YzJoaGNtVndiMmx1ZEM1amIyMThZaUUxZURVeVpraGxNVU5GTmtaR04zbERUMU4yTWxodVdUQldVMkpaY1Zod1RYSm9kV3RXWm1wTlVsUlVRVGhmVlhCbU9GTjNVbkZ5TkUxMllXdEthSFpFZkRBeFZsVkxWelZXU1RJMVJsaFBNalkyUlZkQ1FqTTFRVmhKVTBkRFVVcFdXa1klM0QlMjIlMkMlMjJpJTIyJTNBJTIyNzRmNzM1ZmUtYzg4Ny00MjhhLWFkZmYtNTEyZTg2YmNmZmQzJTIyJTdE) \n",
    "(**Code execution** and **Input, output and options**) on loop for more details)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "def extract_presence_absence(blast_output_tsv: Path, \n",
    "                             tsv_header: str = \"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                             hits_as_string: bool = True,  \n",
    "                             include_match_stats = False, \n",
    "                             pident_threshold: float = 90, \n",
    "                             plen_threshold: float = 60, \n",
    "                             gene_names: list = None\n",
    "                             ) -> dict:\n",
    "    \"\"\"\n",
    "    Parse blast output tsv for best matching alleles\n",
    "    returns:\n",
    "    if include_match stats:\n",
    "        { <gene_name_1>: <allele_number>,  <gene_name_2>: <allele_number>,  <gene_name_3>: <allele_number> ...}\n",
    "    else:\n",
    "        a dictionary (allele_dict) in the format { <gene_name_1>: <allele_number>,  <gene_name_2>: <allele_number>,  <gene_name_3>: <allele_number> ...}\n",
    "\n",
    "    \"\"\"\n",
    "    if os.path.exists(blast_output_tsv):\n",
    "        try:\n",
    "            blast_df = pandas.read_csv(blast_output_tsv, sep='\\t', header = None)\n",
    "            header_list = tsv_header.split(' ')\n",
    "            if len(header_list) == len(blast_df.columns):\n",
    "                blast_df.columns = tsv_header.split(' ')\n",
    "                blast_df[\"plen\"] = blast_df[\"length\"]/blast_df[\"qlen\"]*100\n",
    "                blast_df_unique = blast_df.sort_values(by=['bitscore'], ascending= False).groupby(\"qseqid\").first()\n",
    "                blast_df_filtered = blast_df_unique.query(\n",
    "                            \"plen > @plen_threshold and pident > @pident_threshold\"\n",
    "                        )\n",
    "                blast_dict = dict(blast_df_filtered.to_dict(orient = \"index\"))\n",
    "            else:\n",
    "                print(f\"Failed to parse {blast_output_tsv}. Number of columns do not match length of provided header string\", file=sys.stderr)\n",
    "                return None\n",
    "            \n",
    "        except pandas.errors.EmptyDataError:\n",
    "            blast_dict = {}\n",
    "            print(f\"Blast output file {blast_output_tsv} empty. Assuming 0 blast hits.\")\n",
    "        #except Exception as e:\n",
    "        #    print(f\"Error parsing blast: e\")\n",
    "        if hits_as_string:\n",
    "            \n",
    "            results = []\n",
    "            for gene, d in blast_dict.items():\n",
    "                if include_match_stats:\n",
    "                    results.append(f\"{gene}__{d['pident']}__{d['plen']}\")\n",
    "                else:\n",
    "                    results.append(gene)\n",
    "            result_dict = { \"genes_found\": \", \".join(results)}\n",
    "            return result_dict\n",
    "\n",
    "        else:\n",
    "            result_dict = {}\n",
    "            if gene_names is None:\n",
    "                gene_names = list(blast_dict.keys())\n",
    "            for gene in gene_names:\n",
    "                if gene in blast_dict:\n",
    "                    if include_match_stats:\n",
    "                        result_dict[gene] = f\"{blast_dict[gene]['pident']}__{blast_dict[gene]['plen']}\"\n",
    "                    else:\n",
    "                        result_dict[gene] = \"1\"\n",
    "                else:\n",
    "                    result_dict[gene] = \"0\"\n",
    "            return result_dict\n",
    "\n",
    "\n",
    "\n",
    "    else:\n",
    "        print(f\"No blast output found at {blast_output_tsv}\", file=sys.stderr)\n",
    "        return None\n",
    "    \n",
    "\n",
    "def extract_allele_matches(blast_output_tsv: Path,\n",
    "                           tsv_header: str,\n",
    "                           include_match_stats = False) -> dict:\n",
    "    \"\"\"\n",
    "    Parse blast output tsv for best matching alleles\n",
    "    returns:\n",
    "    if include_match stats:\n",
    "        { <gene_name_1>: <allele_number>__<pident>__<plen>,  <gene_name_2>: <allele_number>__<pident>__<plen>,  <gene_name_3>: <allele_number>__<pident>__<plen> ...}\n",
    "    else:\n",
    "        a dictionary (allele_dict) in the format { <gene_name_1>: <allele_number>,  <gene_name_2>: <allele_number>,  <gene_name_3>: <allele_number> ...}\n",
    "\n",
    "    \"\"\"\n",
    "    allele_dict = {}\n",
    "    detailed_dict = {}\n",
    "    if os.path.exists(blast_output_tsv):\n",
    "        try:\n",
    "            blast_df = pandas.read_csv(blast_output_tsv, sep='\\t', header = None)\n",
    "            header_list = tsv_header.split(' ')\n",
    "            if len(header_list) == len(blast_df.columns):\n",
    "                blast_df.columns = tsv_header.split(' ')\n",
    "                blast_df.set_index(\"qseqid\", drop=False)\n",
    "                blast_df[\"plen\"] = blast_df[\"length\"]/blast_df[\"qlen\"]*100\n",
    "                blast_df[['gene','allele']] = blast_df['qseqid'].str.split('_',expand=True)\n",
    "                blast_df_unique = blast_df.sort_values(by=['bitscore'], ascending= False).groupby(\"gene\").first()\n",
    "                for gene, d in blast_df_unique.to_dict(orient = \"index\").items():\n",
    "                    allele_dict[gene] = d[\"allele\"]\n",
    "                    detailed_dict[gene] = f\"{d['allele']}__{d['pident']}__{d['plen']}\"\n",
    "            else:\n",
    "                print(f\"Failed to parse {blast_output_tsv}. Number of columns do not match length of provided header string\", file=sys.stderr)\n",
    "                return None\n",
    "\n",
    "        except pandas.errors.EmptyDataError:\n",
    "            detailed_dict = {}\n",
    "            allele_dict = {}\n",
    "            print(f\"Blast output file {blast_output_tsv} empty. Assuming 0 blast hits.\")\n",
    "    else:\n",
    "        print(f\"No blast output found at {blast_output_tsv}\", file=sys.stderr)\n",
    "        return None\n",
    "    \n",
    "    if include_match_stats:\n",
    "        return detailed_dict\n",
    "    else:\n",
    "        return allele_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TESTING\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Blast output file ./test_input/blast_parser/empty_gene_presence_absense_test.tsv empty. Assuming 0 blast hits.\n",
      "Blast output file ./test_input/empty_file.txt empty. Assuming 0 blast hits.\n",
      "Blast output file ./test_input/empty_file.txt empty. Assuming 0 blast hits.\n",
      "Blast output file ./test_input/empty_file.txt empty. Assuming 0 blast hits.\n",
      "Blast output file ./test_input/empty_file.txt empty. Assuming 0 blast hits.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No blast output found at ./test_input/file/that/does/not/exist.tsv\n",
      "Failed to parse ./test_input/Legionella/test.sbt.tsv. Number of columns do not match length of provided header string\n",
      "No blast output found at ./test_input/file/that/does/not/exist.tsv\n",
      "Failed to parse ./test_input/Legionella/test.sbt.tsv. Number of columns do not match length of provided header string\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "gene_presence_dict = extract_presence_absence(blast_output_tsv=\"./test_input/blast_parser/gene_presence_absence_test.tsv\",\n",
    "                                     tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                     include_match_stats=False,\n",
    "                                     hits_as_string=True\n",
    "                                     )\n",
    "assert(len(gene_presence_dict) == 1)\n",
    "assert(gene_presence_dict[\"genes_found\"] == \"asd, proA\")\n",
    "\n",
    "\n",
    "gene_presence_dict = extract_presence_absence(blast_output_tsv=\"./test_input/blast_parser/gene_presence_absence_test.tsv\",\n",
    "                                     tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                     include_match_stats=True,\n",
    "                                     hits_as_string=True,\n",
    "                                     )\n",
    "assert(gene_presence_dict[\"genes_found\"] == \"asd__98.732__100.0, proA__97.284__100.0\")\n",
    "\n",
    "\n",
    "gene_presence_dict = extract_presence_absence(blast_output_tsv=\"./test_input/blast_parser/gene_presence_absence_test.tsv\",\n",
    "                                     tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                     include_match_stats=True,\n",
    "                                     hits_as_string=True,\n",
    "                                     pident_threshold=80,\n",
    "                                     )\n",
    "assert(gene_presence_dict[\"genes_found\"] == \"asd__98.732__100.0, pilE__82.835__100.0, proA__97.284__100.0\")\n",
    "\n",
    "\n",
    "gene_presence_dict = extract_presence_absence(blast_output_tsv=\"./test_input/blast_parser/gene_presence_absence_test.tsv\",\n",
    "                                     tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                     include_match_stats=False,\n",
    "                                     hits_as_string=False,\n",
    "                                     gene_names = [\"asd\",\"proA\",\"pilE\"])\n",
    "assert(gene_presence_dict[\"asd\"] == \"1\")\n",
    "assert(gene_presence_dict[\"proA\"] == \"1\")\n",
    "assert(gene_presence_dict[\"pilE\"] == \"0\")\n",
    "\n",
    "\n",
    "gene_presence_dict = extract_presence_absence(blast_output_tsv=\"./test_input/blast_parser/gene_presence_absence_test.tsv\",\n",
    "                                     tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                     include_match_stats=True,\n",
    "                                     hits_as_string=False,\n",
    "                                     gene_names = [\"asd\",\"proA\",\"pilE\"])\n",
    "\n",
    "assert(gene_presence_dict[\"asd\"] == \"98.732__100.0\")\n",
    "assert(gene_presence_dict[\"proA\"] == \"97.284__100.0\")\n",
    "assert(gene_presence_dict[\"pilE\"] == \"0\")\n",
    "\n",
    "\n",
    "gene_presence_dict = extract_presence_absence(blast_output_tsv=\"./test_input/blast_parser/empty_gene_presence_absense_test.tsv\",\n",
    "                                     tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                     include_match_stats=False,\n",
    "                                     hits_as_string=False,\n",
    "                                     gene_names=[\"lag-1\"]\n",
    "                                     )\n",
    "assert(gene_presence_dict[\"lag-1\"] == \"0\")\n",
    "\n",
    "\n",
    "gene_presence_dict = extract_presence_absence(blast_output_tsv=\"./test_input/empty_file.txt\",\n",
    "                                     tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                     include_match_stats=False,\n",
    "                                     hits_as_string=False,\n",
    "                                     gene_names=[\"lag-1\"]\n",
    "                                     )\n",
    "assert(gene_presence_dict[\"lag-1\"] == \"0\")\n",
    "\n",
    "\n",
    "gene_presence_dict = extract_presence_absence(blast_output_tsv=\"./test_input/empty_file.txt\",\n",
    "                                     tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                     include_match_stats=True,\n",
    "                                     hits_as_string=False,\n",
    "                                     gene_names=[\"lag-1\"]\n",
    "                                     )\n",
    "assert(gene_presence_dict[\"lag-1\"] == \"0\")\n",
    "\n",
    "\n",
    "gene_presence_dict = extract_presence_absence(blast_output_tsv=\"./test_input/empty_file.txt\",\n",
    "                                     tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                     include_match_stats=False,\n",
    "                                     hits_as_string=True,\n",
    "                                     gene_names=[\"lag-1\"]\n",
    "                                     )\n",
    "assert(gene_presence_dict[\"genes_found\"] == \"\")\n",
    "\n",
    "\n",
    "gene_presence_dict = extract_presence_absence(blast_output_tsv=\"./test_input/empty_file.txt\",\n",
    "                                     tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                     include_match_stats=True,\n",
    "                                     hits_as_string=True\n",
    "                                     )\n",
    "assert(gene_presence_dict[\"genes_found\"] == \"\")\n",
    "\n",
    "\n",
    "# Test for missing file handling (should return None)\n",
    "gene_presence_dict = extract_presence_absence(blast_output_tsv=\"./test_input/file/that/does/not/exist.tsv\",\n",
    "                                     tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                     include_match_stats=True,\n",
    "                                     hits_as_string=True,\n",
    "                                     gene_names=[\"lag-1\"]\n",
    "                                     )\n",
    "assert(gene_presence_dict is None)\n",
    "\n",
    "# Check for incorrect number of columns handling (should return None)\n",
    "gene_presence_dict = extract_presence_absence(blast_output_tsv=\"./test_input/Legionella/test.sbt.tsv\",\n",
    "                                     tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                     include_match_stats=True,\n",
    "                                     hits_as_string=True,\n",
    "                                     gene_names=[\"lag-1\"]\n",
    "                                     )\n",
    "assert(gene_presence_dict is None)\n",
    "\n",
    "\n",
    "allele_dict = extract_allele_matches(blast_output_tsv=\"./test_input/blast_parser/allele_matches_test.tsv\",\n",
    "                                     tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\")\n",
    "\n",
    "detailed_allele_dict = extract_allele_matches(blast_output_tsv=\"./test_input/blast_parser/allele_matches_test.tsv\",\n",
    "                                     tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                     include_match_stats=True)\n",
    "\n",
    "assert(len(allele_dict) == 7)\n",
    "assert(allele_dict['asd'] == \"9\")\n",
    "assert(detailed_allele_dict['pilE'] == \"3__100.0__100.0\")\n",
    "\n",
    "\n",
    "# Test for missing file handling (should return None)\n",
    "allele_dict = extract_allele_matches(blast_output_tsv=\"./test_input/file/that/does/not/exist.tsv\",\n",
    "                                            tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                            include_match_stats=True)\n",
    "assert(allele_dict is None)\n",
    "\n",
    "\n",
    "# Test for incorrect number of columns handling (should return None)\n",
    "allele_dict = extract_allele_matches(blast_output_tsv=\"./test_input/Legionella/test.sbt.tsv\",\n",
    "                                            tsv_header=\"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",\n",
    "                                            include_match_stats=True)\n",
    "assert(allele_dict is None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# |export\n",
    "from fastcore.script import call_parse\n",
    "\n",
    "\n",
    "@call_parse\n",
    "def presence_absence(\n",
    "    blast_output: Path = None,  # Path to blast output file. Generated with --outfmt 6 option\n",
    "    blast_tsv_header: str = \"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",  # headers in blast output\n",
    "    hits_as_string: bool = True,  # True to print a comma separated list of found genes on a single line. False to return a key: value pair for each gene\n",
    "    include_match_stats: bool = False, # True to include percent identity and percent length in output, false to only include present/absent\n",
    "    percent_identityt: float = 90,  # percent identity threshold for considering a gene present\n",
    "    percent_length: float = 60,  # percent length threshold for considering a gene present\n",
    "    gene_names: list = None,  # name of genes to look for when hits_as_string = False\n",
    "    output_file: Path = None,\n",
    "    config_file: str = None,  # config file to set env vars from\n",
    ") -> None:\n",
    "    \"\"\"\n",
    "    \n",
    "    \"\"\"\n",
    "    #config = core.get_config(config_file)  # Set env vars and get config variables\n",
    "    gene_presence_dict = extract_presence_absence(blast_output_tsv=blast_output,\n",
    "                                                  tsv_header=blast_tsv_header,\n",
    "                                                  hits_as_string=hits_as_string,\n",
    "                                                  include_match_stats=include_match_stats,\n",
    "                                                  pident_threshold=percent_identityt,\n",
    "                                                  plen_threshold=percent_length,\n",
    "                                                  gene_names=gene_names,\n",
    "                                                  )\n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "@call_parse\n",
    "def allele_matches(\n",
    "    blast_output: Path = None,  # Path to blast output file. Generated with --outfmt 6 option\n",
    "    blast_tsv_header: str = \"qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore\",  # headers in blast output\n",
    "    include_match_stats: bool = False, # True to include percent identity and percent length in output, false to only include allele number\n",
    "    output_file: Path = None,\n",
    "    config_file: str = None,  # config file to set env vars from\n",
    ") -> None:\n",
    "    \"\"\"\n",
    "    \n",
    "    \"\"\"\n",
    "    #config = core.get_config(config_file)  # Set env vars and get config variables\n",
    "    allele_dict = extract_allele_matches(blast_output_tsv=blast_output,\n",
    "                                         tsv_header=blast_tsv_header,\n",
    "                                         include_match_stats=include_match_stats,\n",
    "                                         output_file=None,\n",
    "                                         )\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Directive for ensuring that the code in your notebook get executed as a script\n",
    "\n",
    "The code-block here below is required to ensure that the code in the notebook is also transferred to the module (script), otherwise it will just be a notebook. See [Coding in NBdev](https://dksund.sharepoint.com/:fl:/g/contentstorage/CSP_7c761ee7-b577-4e08-8517-bc82392bf65e/ETlSfUyArSNJhX8veMI_JQ8By1aXGHzDJkhotpfpXx4mmw?e=037EwH&nav=cz0lMkZjb250ZW50c3RvcmFnZSUyRkNTUF83Yzc2MWVlNy1iNTc3LTRlMDgtODUxNy1iYzgyMzkyYmY2NWUmZD1iJTIxNXg1MmZIZTFDRTZGRjd5Q09TdjJYblkwVlNiWXFYcE1yaHVrVmZqTVJUVEE4X1VwZjhTd1JxcjRNdmFrSmh2RCZmPTAxVlVLVzVWSlpLSjZVWkFGTkVORVlLN1pQUERCRDZKSVAmYz0lMkYmYT1Mb29wQXBwJnA9JTQwZmx1aWR4JTJGbG9vcC1wYWdlLWNvbnRhaW5lciZ4PSU3QiUyMnclMjIlM0ElMjJUMFJUVUh4a2EzTjFibVF1YzJoaGNtVndiMmx1ZEM1amIyMThZaUUxZURVeVpraGxNVU5GTmtaR04zbERUMU4yTWxodVdUQldVMkpaY1Zod1RYSm9kV3RXWm1wTlVsUlVRVGhmVlhCbU9GTjNVbkZ5TkUxMllXdEthSFpFZkRBeFZsVkxWelZXU1RJMVJsaFBNalkyUlZkQ1FqTTFRVmhKVTBkRFVVcFdXa1klM0QlMjIlMkMlMjJpJTIyJTNBJTIyNzRmNzM1ZmUtYzg4Ny00MjhhLWFkZmYtNTEyZTg2YmNmZmQzJTIyJTdE) \n",
    "(**Writing your own notebooks**) on loop for more details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | hide\n",
    "# This is included at the end to ensure when you run through your notebook the code is also transferred to the module and isn't just a notebook\n",
    "import nbdev\n",
    "\n",
    "nbdev.nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
